pkg/agent/client/dial.go

```diff
@@ -38,23 +51,46 @@ func DialServer(ctx context.Context, config DialServerConfig) (*grpc.ClientConn,
 		// trust domain. The peer certificate must present the server SPIFFE
 		// ID for the trust domain.
 		VerifyPeerCertificate: func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
-			roots := x509.NewCertPool()
-			for _, c := range config.GetBundle() {
-				roots.AddCert(c)
-			}
-			trustDomainRoots := map[string]*x509.CertPool{
-				idutil.TrustDomainID(config.TrustDomain): roots,
-			}
 			var serverChain []*x509.Certificate
 			for _, rawCert := range rawCerts {
 				cert, err := x509.ParseCertificate(rawCert)
 				if err != nil {
-					return err
+					return errs.Wrap(err)
 				}
 				serverChain = append(serverChain, cert)
 			}
 
-			_, err := spiffe.VerifyPeerCertificate(serverChain, trustDomainRoots, spiffe.ExpectPeer(idutil.ServerID(config.TrustDomain)))
+			if len(serverChain) == 0 {
+				// This is not really possible without a catastrophic bug
+				// creeping into the TLS stack.
+				return errs.New("server chain is unexpectedly empty")
+			}
+
+			expectedServerID := idutil.ServerID(config.TrustDomain)
+
+			if config.GetBundle == nil {
```

Comment:
Similar to my above comment, we should probably check InsecureBootstrap directly for clarity's sake
```suggestion
			if config.InsecureBootstrap {
```

Reply:
Checking only for InsecureBootstrap is insufficient since once we have a bundle available, we want to use it for future connections, correct?